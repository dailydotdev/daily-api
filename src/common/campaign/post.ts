import { ValidationError } from 'apollo-server-errors';
import { AuthContext } from '../../Context';
import {
  ArticlePost,
  CampaignPost,
  CampaignState,
  CampaignType,
  Post,
  PostType,
  type ConnectionManager,
  type FreeformPost,
  type SharePost,
} from '../../entity';
import { getPostPermalink } from '../../schema/posts';
import {
  type GetCampaignListResponse,
  type GetCampaignResponse,
} from '../../integrations/skadi';
import type { Connection } from 'graphql-relay';
import { In } from 'typeorm';
import { mapCloudinaryUrl } from '../cloudinary';
import { pickImageUrl } from '../post';
import { NotFoundError } from '../../errors';
import { debeziumTimeToDate, systemUser, updateFlagsStatement } from '../utils';
import { getDiscussionLink } from '../links';
import { skadiApiClientV1 } from '../../integrations/skadi/api/v1/clients';
import { largeNumberFormat } from '../devcard';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import { truncatePostToTweet } from '../twitter';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { usdToCores } from '../number';

import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import { randomUUID } from 'crypto';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { addDays } from 'date-fns';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';

export interface GQLPromotedPost
  extends Omit<
    GetCampaignResponse,
    'spend' | 'budget' | 'startedAt' | 'endedAt'
  > {
  spend: number;
  budget: number;
  startedAt: Date;
  endedAt: Date;
}

export const validatePostBoostPermissions = async (
  ctx: AuthContext,
  postId: string,
): Promise<Pick<Post, 'id' | 'flags'>> => {
  const { userId } = ctx;

  return ctx.con.getRepository(Post).findOneOrFail({
    select: ['id', 'flags'],
    where: [
      { id: postId, authorId: userId },
      { id: postId, scoutId: userId },
    ],
  });
};

export const checkPostAlreadyBoosted = (post: Pick<Post, 'flags'>): void => {
  if (!!post.flags?.campaignId) {
    throw new ValidationError('Post is already boosted');
  }
};

interface CampaignBoostedPost
  extends Pick<Post, 'id' | 'shortId' | 'title' | 'slug'> {
  image: string;
  permalink: string;
  engagements: number;
  commentsPermalink?: string;
}

export interface GQLBoostedPost {
  campaign: GQLPromotedPost;
  post: CampaignBoostedPost;
}

interface GetBoostedPost extends CampaignBoostedPost {
  type: PostType;
  sharedTitle?: string;
  sharedImage?: string;
  views: number;
  upvotes: number;
  comments: number;
}

const getBoostedPostBuilder = (con: ConnectionManager, alias = 'p1') =>
  con
    .getRepository(Post)
    .createQueryBuilder(alias)
    .select(`"${alias}".id`, 'id')
    .addSelect(`"${alias}"."shortId"`, 'shortId')
    .addSelect(`"${alias}".slug`, 'slug')
    .addSelect(`"${alias}".image`, 'image')
    .addSelect(`"${alias}".title`, 'title')
    .addSelect(`"${alias}".type`, 'type')
    .addSelect(`"${alias}".upvotes::int`, 'upvotes')
    .addSelect(`"${alias}".comments::int`, 'comments')
    .addSelect(`"${alias}".views::int`, 'views')
    .addSelect('p2.title', 'sharedTitle')
    .addSelect('p2.image', 'sharedImage')
    .leftJoin(Post, 'p2', `"${alias}"."sharedPostId" = p2.id`);

export const getBoostedPost = async (
  con: ConnectionManager,
  id: string,
): Promise<GetBoostedPost> => {
  const result = await getBoostedPostBuilder(con).where({ id }).getRawOne();

  if (!result) {
    throw new NotFoundError('Post does not exist');
  }

  return result;
};

export const getFormattedBoostedPost = (
  post: GetBoostedPost,
): GQLBoostedPost['post'] => {
  const { id, shortId, sharedImage, sharedTitle, slug } = post;
  let image: string | undefined = post.image;
  let title = post.title;

  if (post.type === PostType.Share) {
    image = sharedImage;
    title = title || sharedTitle;
  }

  return {
    id,
    slug,
    shortId,
    title,
    image: mapCloudinaryUrl(image) ?? pickImageUrl({ createdAt: new Date() }),
    permalink: getPostPermalink({ shortId }),
    commentsPermalink: post.slug ? getDiscussionLink(post.slug) : undefined,
    engagements: post.comments + post.upvotes + post.views,
  };
};

export const getFormattedCampaign = ({
  spend,
  budget,
  startedAt,
  endedAt,
  ...campaign
}: GetCampaignResponse): GQLPromotedPost => ({
  ...campaign,
  spend: usdToCores(parseFloat(spend)),
  budget: usdToCores(parseFloat(budget)),
  startedAt: debeziumTimeToDate(startedAt),
  endedAt: debeziumTimeToDate(endedAt),
});

export interface BoostedPostStats
  extends Pick<GetCampaignListResponse, 'clicks' | 'impressions' | 'users'> {
  engagements: number;
  totalSpend: number;
}

export interface BoostedPostConnection extends Connection<GQLBoostedPost> {
  stats?: BoostedPostStats;
}

export const consolidateCampaignsWithPosts = async (
  campaigns: GetCampaignResponse[],
  con: ConnectionManager,
): Promise<GQLBoostedPost[]> => {
  const ids = campaigns.map(({ postId }) => postId);
  const builder = getBoostedPostBuilder(con);
  const postAlias = 'p1';
  const posts = await builder
    .where(`"${postAlias}".id IN (:...ids)`, { ids })
    .getRawMany<GetBoostedPost>();
  const mapped = posts.reduce(
    (map, post) => ({ ...map, [post.id]: post }),
    {} as Record<string, GetBoostedPost>,
  );

  return campaigns.map((campaign) => ({
    campaign: getFormattedCampaign(campaign),
    post: getFormattedBoostedPost(mapped[campaign.postId]),
  }));
};

interface TotalEngagements
  extends Pick<Post, 'views' | 'comments' | 'upvotes'> {}

export const getTotalEngagements = async (
  con: ConnectionManager,
  postIds: string[],
): Promise<number> => {
  const builder = con.getRepository(Post).createQueryBuilder();

  const engagements = await builder
    .select('SUM(upvotes)::int', 'upvotes')
    .addSelect('SUM(comments)::int', 'comments')
    .addSelect('SUM(views)::int', 'views')
    .where({ id: In(postIds) })
    .getRawOne<TotalEngagements>();

  return (
    (engagements?.upvotes || 0) +
    (engagements?.comments || 0) +
    (engagements?.views || 0)
  );
};

export const generateBoostEmailUpdate: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await skadiApiClientV1.getCampaignById({
    campaignId: notification.referenceId!,
    userId: user.id,
  });

  if (!campaign) {
    return null;
  }

  const post = await con.getRepository(Post).findOne({
    where: { id: campaign.postId },
  });

  if (!post) {
    return null;
  }

  const sharedPost = await (post.type === PostType.Share
    ? con.getRepository(ArticlePost).findOne({
        where: { id: (post as SharePost).sharedPostId },
        select: ['title', 'image', 'slug'],
      })
    : Promise.resolve(null));

  const title = truncatePostToTweet(post || sharedPost);
  const engagement = post.views + post.upvotes + post.comments;

  return {
    start_date: formatMailDate(debeziumTimeToDate(campaign.startedAt)),
    end_date: formatMailDate(debeziumTimeToDate(campaign.endedAt)),
    impressions: largeNumberFormat(campaign.impressions),
    clicks: largeNumberFormat(campaign.clicks),
    engagement: largeNumberFormat(engagement),
    post_link: getDiscussionLink(post.slug),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    post_image: sharedPost?.image || (post as FreeformPost).image,
    post_title: title,
  };
};

export const getAdjustedReach = (value: number) => {
  // We do plus-minus 8% of the generated value
  const difference = Math.floor(value * 0.08);
  const min = Math.max(value - difference, 0);
  const estimatedReach = {
    min,
    max: Math.max(value + difference, min),
  };

  return estimatedReach;
};

export const startCampaignPost = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value: postId } = args;
  const post = await validatePostBoostPermissions(ctx, postId);
  checkPostAlreadyBoosted(post);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const creativeId = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignPost).save(
      manager.getRepository(CampaignPost).create({
        id,
        creativeId,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: postId,
        state: CampaignState.Active,
        endedAt,
        postId,
        type: CampaignType.Post,
      }),
    );

    const campaignId = campaign.id;
    const tags = await getPostTags(manager, postId);

    await skadiApiClientV2.startCampaign(campaign, tags);

    await manager
      .getRepository(Post)
      .update(
        { id: postId },
        { flags: updateFlagsStatement<Post>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Post Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignPost = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(Post)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Post>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Post Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const getPostTags = async (con: ConnectionManager, postId: string) => {
  const builder = con.getRepository(Post).createQueryBuilder('p1');
  const subquery = builder
    .subQuery()
    .select(`COALESCE(p2."tagsStr", '')`)
    .from(Post, 'p2')
    .where('p2.id = p1."sharedPostId"')
    .getQuery();

  const result = await builder
    .select(`COALESCE(p1."tagsStr", '')`, 'tagsStr')
    .addSelect(`(${subquery})`, 'sharedTagsStr')
    .where('p1.id = :id', { id: postId })
    .getRawOne<{ tagsStr: string; sharedTagsStr: string }>();

  const tags1 = (result?.tagsStr ?? '').split(',');
  const tags2 = (result?.sharedTagsStr ?? '').split(',');
  const list = [...tags1, ...tags2].filter((tag) => tag.trim().length > 0);

  return Array.from(new Set(list));
};
