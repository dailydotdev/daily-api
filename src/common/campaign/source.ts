import {
  CampaignSource,
  CampaignState,
  CampaignType,
  Post,
  Source,
  type ConnectionManager,
} from '../../entity';

import type { AuthContext } from '../../Context';
import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import {
  ensureSourcePermissions,
  SourcePermissions,
} from '../../schema/sources';
import { ValidationError } from 'apollo-server-errors';
import { randomUUID } from 'crypto';
import { addDays } from 'date-fns';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { usdToCores } from '../number';
import { systemUser, updateFlagsStatement } from '../utils';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationBuilder } from '../../notifications/builder';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import type { NotificationCampaignSourceContext } from '../../notifications';

export const validateSquadBoostPermissions = async (
  ctx: AuthContext,
  sourceId: string,
) => {
  const source = await ensureSourcePermissions(
    ctx,
    sourceId,
    SourcePermissions.BoostSquad,
  );

  if (source.flags.campaignId) {
    throw new ValidationError('Source already has a campaign');
  }

  return source;
};

export const getSourceTags = async (
  con: ConnectionManager,
  sourceId: string,
): Promise<string[]> => {
  const result = await con.getRepository(Post).query<{ tag: string }[]>(
    `
      WITH recent_posts AS (
        SELECT id, "sharedPostId"
        FROM post
        WHERE "sourceId" = $1
        ORDER BY "createdAt" DESC
        LIMIT 30
      )
      SELECT DISTINCT ps.keyword AS tag
      FROM post_keyword ps
      INNER JOIN recent_posts rp
        ON ps."postId" = rp.id
        OR ps."postId" = rp."sharedPostId"
      WHERE ps.status = 'allow'
      LIMIT 30;
    `,
    [sourceId],
  );

  return result.map(({ tag }) => tag.trim()).filter(Boolean);
};

export const startCampaignSource = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value } = args;
  const source = await validateSquadBoostPermissions(ctx, value);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignSource).save(
      manager.getRepository(CampaignSource).create({
        id,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: source.id,
        state: CampaignState.Active,
        endedAt,
        sourceId: source.id,
        type: CampaignType.Squad,
      }),
    );

    const campaignId = campaign.id;
    const last30tags = await getSourceTags(manager, source.id);
    const finalTags = last30tags.length > 3 ? last30tags : []; // when it is 3 or below, we set global targeting

    await skadiApiClientV2.startCampaign(campaign, finalTags);

    await manager
      .getRepository(Source)
      .update(
        { id: source.id },
        { flags: updateFlagsStatement<Source>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Squad Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignSource = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(CampaignSource)
      .update({ id: campaignId }, { state: CampaignState.Cancelled });

    await manager
      .getRepository(Source)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Source>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Squad Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const generateCampaignSquadNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignSourceContext,
) => {
  const { campaign, source, event, user } = ctx;

  return builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`)
    .avatarSource(source);
};

export const generateCampaignSquadEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(CampaignSource)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  const source = await con
    .getRepository(Source)
    .findOneByOrFail({ id: campaign.sourceId });

  return {
    start_date: formatMailDate(campaign.createdAt),
    end_date: formatMailDate(campaign.endedAt),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    source_image: source.image,
    source_handle: source.handle,
    source_name: source.name,
  };
};
