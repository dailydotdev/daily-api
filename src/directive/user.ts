import { DataSource } from 'typeorm';
import { FastifyLoggerInstance } from 'fastify';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';

export const deleteUser = async (
  con: DataSource,
  logger: FastifyLoggerInstance,
  userId: string,
  messageId?: string,
) => {
  try {
    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: '404',
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: '404' });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(User).delete(userId);
    });
    if (logger) {
      logger.info(
        {
          userId,
          messageId,
        },
        'deleted user',
      );
    }
  } catch (err) {
    if (logger) {
      logger.error(
        {
          userId,
          messageId,
          err,
        },
        'failed to delete user',
      );
    }
    throw err;
  }
};
