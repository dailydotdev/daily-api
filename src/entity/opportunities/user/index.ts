export * from './OpportunityUser';
export * from './OpportunityUserRecruiter';
