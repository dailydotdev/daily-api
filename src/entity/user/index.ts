export * from './User';
export * from './UserPost';
export * from './UserState';
export * from './UserAction';
export * from './UserStreak';
export * from './UserPersonalizedDigest';
export * from './UserMarketingCta';
