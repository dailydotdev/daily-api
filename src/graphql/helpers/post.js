const getFeedParams =
  ({ user, post }, args, rankBy, filters = {}, ignoreUserFilters = false, ip) => {
    const userId = user ? user.userId : null;

    return {
      fields: userId ? [...post.defaultUserFields, 'read'] : post.defaultAnonymousFields,
      filters: Object.assign({}, { before: args.latest }, filters),
      rankBy,
      userId,
      ignoreUserFilters,
      page: args.page,
      pageSize: args.pageSize,
      ip,
    };
  };

const assignType = type => x => Object.assign({}, x, { type });

export default {
  getFeedParams,
  assignType,
};
