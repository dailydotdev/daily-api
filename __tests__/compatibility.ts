import {
  disposeGraphQLTesting,
  GraphQLTestClient,
  GraphQLTestingState,
  initializeGraphQLTesting,
  MockContext,
  saveFixtures,
} from './helpers';
import {
  ArticlePost,
  Bookmark,
  Post,
  PostKeyword,
  Source,
  User,
} from '../src/entity';
import { sourcesFixture } from './fixture/source';
import { postKeywordsFixture, postsFixture } from './fixture/post';
import { DataSource } from 'typeorm';
import createOrGetConnection from '../src/db';
import { usersFixture } from './fixture/user';

let con: DataSource;
let state: GraphQLTestingState;
let client: GraphQLTestClient;
let loggedUser: string = null;

beforeAll(async () => {
  con = await createOrGetConnection();
  state = await initializeGraphQLTesting(
    () => new MockContext(con, loggedUser),
  );
  client = state.client;
});

beforeEach(async () => {
  loggedUser = null;

  await saveFixtures(con, Source, sourcesFixture);
  await saveFixtures(con, ArticlePost, postsFixture);
  await saveFixtures(con, PostKeyword, postKeywordsFixture);
});

afterAll(() => disposeGraphQLTesting(state));

const now = new Date();
const bookmarksFixture = [
  {
    userId: '1',
    postId: 'p3',
    createdAt: new Date(now.getTime() - 1000),
  },
  {
    userId: '1',
    postId: 'p1',
    createdAt: new Date(now.getTime() - 2000),
  },
  {
    userId: '1',
    postId: 'p5',
    createdAt: new Date(now.getTime() - 3000),
  },
];

const feedFields = `
    id
    url
    title
    readTime
    tags
    publication {
      id
      name
      image
    }
`;

describe('query latest', () => {
  const QUERY = `
  query Latest($params: QueryPostInput) {
    latest(params: $params) {
      ${feedFields}
    }
  }`;

  it('should return anonymous feed with no filters ordered by popularity', async () => {
    await con.getRepository(Post).delete({ id: 'p6' });
    const latest = new Date().toISOString();
    const res = await client.query(QUERY, {
      variables: { params: { latest } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should return anonymous feed with no filters ordered by time', async () => {
    await con.getRepository(Post).delete({ id: 'p6' });
    const latest = new Date().toISOString();
    const res = await client.query(QUERY, {
      variables: { params: { latest, sortBy: 'creation' } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should return anonymous feed filtered by sources', async () => {
    const latest = new Date().toISOString();
    const res = await client.query(QUERY, {
      variables: { params: { latest, pubs: 'a,b' } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should return anonymous feed filtered by tags', async () => {
    const latest = new Date().toISOString();
    const res = await client.query(QUERY, {
      variables: { params: { latest, tags: 'html,webdev' } },
    });
    expect(res.data).toMatchSnapshot();
  });
});

describe('query bookmarks', () => {
  const QUERY = `
  query Bookmarks($params: QueryPostInput) {
    bookmarks(params: $params) {
      ${feedFields}
    }
  }`;

  beforeEach(async () => {
    await saveFixtures(con, User, usersFixture);
  });

  it('should return bookmarks ordered by time', async () => {
    const latest = new Date().toISOString();
    loggedUser = '1';
    await saveFixtures(con, Bookmark, bookmarksFixture);
    const res = await client.query(QUERY, {
      variables: { params: { latest } },
    });
    expect(res.data).toMatchSnapshot();
  });
});

describe('query postsByTag', () => {
  const QUERY = `
  query PostsByTag($params: PostByTagInput) {
    postsByTag(params: $params) {
      ${feedFields}
    }
  }`;

  it('should return a single tag feed', async () => {
    const latest = new Date().toISOString();
    const res = await client.query(QUERY, {
      variables: { params: { latest, tag: 'javascript' } },
    });
    expect(res.data).toMatchSnapshot();
  });
});
