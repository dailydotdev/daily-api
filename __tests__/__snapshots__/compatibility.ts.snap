// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query bookmarks should return bookmarks ordered by time 1`] = `
Object {
  "bookmarks": Array [
    Object {
      "id": "p3",
      "publication": Object {
        "id": "c",
        "image": "http://image.com/c",
        "name": "C",
      },
      "readTime": null,
      "tags": Array [],
      "title": "P3",
      "url": "http://p3.com",
    },
    Object {
      "id": "p1",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "javascript",
        "webdev",
      ],
      "title": "P1",
      "url": "http://p1.com",
    },
    Object {
      "id": "p5",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [
        "html",
        "javascript",
      ],
      "title": "P5",
      "url": "http://p5.com",
    },
  ],
}
`;

exports[`query latest should return anonymous feed filtered by sources 1`] = `
Object {
  "latest": Array [
    Object {
      "id": "p5",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [
        "html",
        "javascript",
      ],
      "title": "P5",
      "url": "http://p5.com",
    },
    Object {
      "id": "p2",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [],
      "title": "P2",
      "url": "http://p2.com",
    },
    Object {
      "id": "p4",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "backend",
        "data",
        "javascript",
      ],
      "title": "P4",
      "url": "http://p4.com",
    },
    Object {
      "id": "p1",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "javascript",
        "webdev",
      ],
      "title": "P1",
      "url": "http://p1.com",
    },
  ],
}
`;

exports[`query latest should return anonymous feed filtered by tags 1`] = `
Object {
  "latest": Array [
    Object {
      "id": "p5",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [
        "html",
        "javascript",
      ],
      "title": "P5",
      "url": "http://p5.com",
    },
    Object {
      "id": "p1",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "javascript",
        "webdev",
      ],
      "title": "P1",
      "url": "http://p1.com",
    },
  ],
}
`;

exports[`query latest should return anonymous feed with no filters ordered by popularity 1`] = `
Object {
  "latest": Array [
    Object {
      "id": "p5",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [
        "html",
        "javascript",
      ],
      "title": "P5",
      "url": "http://p5.com",
    },
    Object {
      "id": "p2",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [],
      "title": "P2",
      "url": "http://p2.com",
    },
    Object {
      "id": "p3",
      "publication": Object {
        "id": "c",
        "image": "http://image.com/c",
        "name": "C",
      },
      "readTime": null,
      "tags": Array [],
      "title": "P3",
      "url": "http://p3.com",
    },
    Object {
      "id": "p4",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "backend",
        "data",
        "javascript",
      ],
      "title": "P4",
      "url": "http://p4.com",
    },
    Object {
      "id": "p1",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "javascript",
        "webdev",
      ],
      "title": "P1",
      "url": "http://p1.com",
    },
    Object {
      "id": "yt2",
      "publication": Object {
        "id": "squad",
        "image": "http//image.com/s",
        "name": "Squad",
      },
      "readTime": null,
      "tags": Array [],
      "title": "youtube post",
      "url": "https://youtu.be/Oso6dYXw5lc",
    },
  ],
}
`;

exports[`query latest should return anonymous feed with no filters ordered by time 1`] = `
Object {
  "latest": Array [
    Object {
      "id": "p1",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "javascript",
        "webdev",
      ],
      "title": "P1",
      "url": "http://p1.com",
    },
    Object {
      "id": "p2",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [],
      "title": "P2",
      "url": "http://p2.com",
    },
    Object {
      "id": "p3",
      "publication": Object {
        "id": "c",
        "image": "http://image.com/c",
        "name": "C",
      },
      "readTime": null,
      "tags": Array [],
      "title": "P3",
      "url": "http://p3.com",
    },
    Object {
      "id": "p4",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "backend",
        "data",
        "javascript",
      ],
      "title": "P4",
      "url": "http://p4.com",
    },
    Object {
      "id": "p5",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [
        "html",
        "javascript",
      ],
      "title": "P5",
      "url": "http://p5.com",
    },
    Object {
      "id": "yt2",
      "publication": Object {
        "id": "squad",
        "image": "http//image.com/s",
        "name": "Squad",
      },
      "readTime": null,
      "tags": Array [],
      "title": "youtube post",
      "url": "https://youtu.be/Oso6dYXw5lc",
    },
  ],
}
`;

exports[`query postsByTag should return a single tag feed 1`] = `
Object {
  "postsByTag": Array [
    Object {
      "id": "p5",
      "publication": Object {
        "id": "b",
        "image": "http://image.com/b",
        "name": "B",
      },
      "readTime": null,
      "tags": Array [
        "html",
        "javascript",
      ],
      "title": "P5",
      "url": "http://p5.com",
    },
    Object {
      "id": "p4",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "backend",
        "data",
        "javascript",
      ],
      "title": "P4",
      "url": "http://p4.com",
    },
    Object {
      "id": "p1",
      "publication": Object {
        "id": "a",
        "image": "http://image.com/a",
        "name": "A",
      },
      "readTime": null,
      "tags": Array [
        "javascript",
        "webdev",
      ],
      "title": "P1",
      "url": "http://p1.com",
    },
  ],
}
`;
